### PR TITLE
#20488 3.0 core dictionary get key

### PR DESCRIPTION
**Pull request by malcolmhoward**
This pull request addresses https://github.com/godotengine/godot/issues/20488.

### CODE REVIEW

**Comment by malcolmhoward** [comment]
`Variant Dictionary::get(const Variant &p_key) const {
	OrderedHashMap<Variant, Variant, VariantHasher, VariantComparator>::ConstElement E = ((const OrderedHashMap<Variant, Variant, VariantHasher, VariantComparator> *)&_p->variant_map)->find(p_key);
	if (!E)
		return Variant();
	return E.get();
}
`

I copied, pasted, and renamed get_valid to make the above method.  I'd prefer to maybe refactor VCALL_LOCALMEM1R or ADDFUNC1R to allow aliasing of the the new .get() method on the existing get_valid() method.

File: core/dictionary.cpp [hunk 1]
```diff
@@ -86,6 +86,15 @@ Variant Dictionary::get_valid(const Variant &p_key) const {
...
 
+Variant Dictionary::get(const Variant &p_key) const {
+
+	OrderedHashMap<Variant, Variant, VariantHasher, VariantComparator>::ConstElement E = ((const OrderedHashMap<Variant, Variant, VariantHasher, VariantComparator> *)&_p->variant_map)->find(p_key);
+
+	if (!E)
+		return Variant();
+	return E.get();
+}
+
```
Comment (neikeq) [inline core/dictionary.cpp]: This can be simplified to:

``` C++
Variant *ret = getptr(p_key);
if (!ret)
    return Variant();
return *ret;
```

**Comment by vnen** [comment]
Not sure why you changed the base branch from `master` to `3.0`. We request that all changes are done in master, and if relevant it can cherry-picked to 3.0 by the stable branch maintainer.

Also, you could've just called the other function if you didn't want to repeat the code.

**Comment by vnen** [comment]
Oops, it seems your branch is still based on 3.0, so you are effectively applying all the 3.0 commits onto master here. You need to [rebase](http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase) your branch on top of master to leave this PR as a single commit.

**Comment by williamd1k0** [comment]
Yo, I was also trying to implement it to do some c++ studies.

I'm not going to open another PR (I don't want to sound arrogant or anything like that), but I'll share the diff here (feel free to review the code):

```diff
diff --git a/core/dictionary.cpp b/core/dictionary.cpp
index 42d9eab31..75a1cf0f0 100644
--- a/core/dictionary.cpp
+++ b/core/dictionary.cpp
@@ -112,6 +112,14 @@ Variant Dictionary::get_valid(const Variant &p_key) const {
 	return E.get();
 }
 
+Variant Dictionary::get(const Variant &p_key, const Variant &p_default) const {
+
+	const Variant *ret = getptr(p_key);
+	if (!ret)
+		return p_default;
+	return *ret;
+}
+
 int Dictionary::size() const {
 
 	return _p->variant_map.size();
diff --git a/core/dictionary.h b/core/dictionary.h
index 00ec67fb9..402d5b5e4 100644
--- a/core/dictionary.h
+++ b/core/dictionary.h
@@ -57,6 +57,7 @@ public:
 	Variant *getptr(const Variant &p_key);
 
 	Variant get_valid(const Variant &p_key) const;
+	Variant get(const Variant &p_key, const Variant &p_default) const;
 
 	int size() const;
 	bool empty() const;
diff --git a/core/variant_call.cpp b/core/variant_call.cpp
index 20a2929dc..95af64563 100644
--- a/core/variant_call.cpp
+++ b/core/variant_call.cpp
@@ -483,6 +483,7 @@ struct _VariantCall {
 	VCALL_LOCALMEM0R(Dictionary, keys);
 	VCALL_LOCALMEM0R(Dictionary, values);
 	VCALL_LOCALMEM1R(Dictionary, duplicate);
+	VCALL_LOCALMEM2R(Dictionary, get);
 
 	VCALL_LOCALMEM2(Array, set);
 	VCALL_LOCALMEM1R(Array, get);
@@ -1654,6 +1655,7 @@ void register_variant_methods() {
 	ADDFUNC0R(DICTIONARY, ARRAY, Dictionary, keys, varray());
 	ADDFUNC0R(DICTIONARY, ARRAY, Dictionary, values, varray());
 	ADDFUNC1R(DICTIONARY, DICTIONARY, Dictionary, duplicate, BOOL, "deep", varray(false));
+	ADDFUNC2R(DICTIONARY, NIL, Dictionary, get, NIL, "key", NIL, "default", varray(Variant()));
 
 	ADDFUNC0R(ARRAY, INT, Array, size, varray());
 	ADDFUNC0R(ARRAY, BOOL, Array, empty, varray());
```

These changes also implement the `default` param, as I'd commented in #20488

**Pull request closed without merging**